source/common/unified_memory_allocation/src/memory_tracker.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+
+#include "memory_tracker_internal.h"
+
+#include <map>
+#include <mutex>
+#include <shared_mutex>
+
+// TODO: reimplement in C and optimize...
+struct uma_memory_tracker_t {
+    enum uma_result_t add(void *pool, const void *ptr, size_t size) {
+        std::unique_lock<std::shared_mutex> lock(mtx);
+
+        auto ret = map.try_emplace(reinterpret_cast<uintptr_t>(ptr), size, pool);
+        return ret.second ? UMA_RESULT_SUCCESS : UMA_RESULT_ERROR_UNKNOWN;
+    }
+
+    enum uma_result_t remove(const void *ptr, size_t size) {
+        std::unique_lock<std::shared_mutex> lock(mtx);
+
+        map.erase(reinterpret_cast<uintptr_t>(ptr));
+
+        // TODO: handle removing part of the range
+        (void)size;
+
+        return UMA_RESULT_SUCCESS;
+    }
+
+    void *find(const void *ptr) {
+        std::shared_lock<std::shared_mutex> lock(mtx);
```

Comment:
I actually just realized std::map might not be threadsafe (even for upper_bound) so I'll need to reimplement this probably.

Reply:
What do you mean? None of this is thread-safe since you are not using a const map. That's why you needed a lock.
And, even if you had used a lock-free map, you'd need a single safe find 'less equals' method.

Reply:
I wasn't sure if `const` version of `upper_bound` is thread safe but it turns out it is - so this code should be safe as long as we make sure we use the const version.

Eventually, we'll need to reimplement this anyway with something faster (ideally with lock-free reads), for example with https://github.com/kilobyte/critnib